# 📘 Industrial Machine Predictive Maintenance

## Data Understanding

### Project Objective

The objective of this notebook is to understand the dataset before performing any preprocessing or model building. This includes exploring the dataset structure, identifying feature types, checking data quality, and understanding the target variable for the regression task.

# 1. Import Required Libraries

In this section, we import the required Python libraries that will be used throughout the notebook for data loading and analysis.

In [1]:
import pandas as pd

import numpy as np

# 2. Load the Dataset

Load the industrial machine predictive maintenance dataset into a Pandas DataFrame for further exploration.

In [22]:
df=pd.read_csv("./data/predictive_maintenance.csv")

# 3. Display the First Five Records

Display the first five rows of the dataset to get an initial understanding of the available features and data format.

In [26]:
df.head()

,timestamp,machine_id,machine_type,vibration_rms,temperature_motor,current_phase_avg,pressure_level,rpm,operating_mode,hours_since_maintenance,ambient_temp,rul_hours,failure_within_24h,failure_type,estimated_repair_cost
0,2024-01-01 00:00:00,1,CNC,0.81,49.51,5.10,23.6,860.9,idle,273.80,13.9,61.00,0,none,0
1,2024-01-01 00:03:00,1,CNC,0.75,40.58,5.30,23.6,899.6,idle,273.85,10.2,60.95,0,none,0
2,2024-01-01 00:21:00,1,CNC,0.71,49.70,NaN,21.3,862.7,idle,274.15,13.6,60.65,0,none,0
3,2024-01-01 00:45:00,1,CNC,0.76,43.04,4.79,22.6,870.4,idle,274.55,13.4,60.25,0,none,0
4,2024-01-01 00:54:00,1,CNC,0.88,41.39,4.44,22.2,881.9,idle,274.70,10.8,60.10,0,none,0


# 4. Check Dataset Dimensions

Determine the total number of rows and columns present in the dataset.

In [23]:
df.shape

(24042, 15)

# 5. Display Feature Names

List all feature names available in the dataset to understand the input variables and identify the target variable.

In [24]:
df.columns

Index(['timestamp', 'machine_id', 'machine_type', 'vibration_rms',
       'temperature_motor', 'current_phase_avg', 'pressure_level', 'rpm',
       'operating_mode', 'hours_since_maintenance', 'ambient_temp',
       'rul_hours', 'failure_within_24h', 'failure_type',
       'estimated_repair_cost'],
      dtype='object')

In [25]:
df['machine_type'].value_counts()

machine_type
Pump           6114
Robotic Arm    6003
Compressor     5988
CNC            5937
Name: count, dtype: int64

# 6. Explore Dataset Information

Inspect the dataset to understand:

- Number of entries
- Data types of each feature
- Memory usage
- Missing values (if any)

In [28]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24042 entries, 0 to 24041
Data columns (total 15 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   timestamp                24042 non-null  object 
 1   machine_id               24042 non-null  int64  
 2   machine_type             24042 non-null  object 
 3   vibration_rms            23042 non-null  float64
 4   temperature_motor        23208 non-null  float64
 5   current_phase_avg        23311 non-null  float64
 6   pressure_level           23118 non-null  float64
 7   rpm                      23509 non-null  float64
 8   operating_mode           24042 non-null  object 
 9   hours_since_maintenance  24042 non-null  float64
 10  ambient_temp             24042 non-null  float64
 11  rul_hours                24042 non-null  float64
 12  failure_within_24h       24042 non-null  int64  
 13  failure_type             24042 non-null  object 
 14  estimated_repair_cost 

In [29]:
print("Rows :", df.shape[0])
print("Columns :", df.shape[1])

Rows : 24042
Columns : 15


In [30]:
df.columns.tolist()

['timestamp',
 'machine_id',
 'machine_type',
 'vibration_rms',
 'temperature_motor',
 'current_phase_avg',
 'pressure_level',
 'rpm',
 'operating_mode',
 'hours_since_maintenance',
 'ambient_temp',
 'rul_hours',
 'failure_within_24h',
 'failure_type',
 'estimated_repair_cost']

# 7. Check Missing Values

Identify whether any feature contains missing values that need to be handled during preprocessing.

In [40]:
df.isnull().sum()

timestamp                     0
machine_id                    0
machine_type                  0
vibration_rms              1000
temperature_motor           834
current_phase_avg           731
pressure_level              924
rpm                         533
operating_mode                0
hours_since_maintenance       0
ambient_temp                  0
rul_hours                     0
failure_within_24h            0
failure_type                  0
estimated_repair_cost         0
dtype: int64

In [41]:
df.isnull().sum().sum()

4022

# 8. Check Duplicate Records

Verify whether duplicate rows exist in the dataset.

In [32]:
df.duplicated().sum()

0

# 9. Generate Statistical Summary

Generate descriptive statistics for all numerical features, including:

- Mean
- Standard Deviation
- Minimum
- Maximum
- Quartiles

In [35]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
machine_id,24042.0,10.505033,5.746455,1.00,6.00,10.00,15.0000,20.00
vibration_rms,23042.0,1.623667,1.081061,0.35,0.82,1.27,2.2700,10.00
temperature_motor,23208.0,51.404295,12.519279,28.00,42.61,50.06,59.9625,95.00
current_phase_avg,23311.0,8.823829,5.366391,2.20,4.63,6.43,13.1200,35.00
pressure_level,23118.0,59.012233,38.723271,10.10,22.70,46.30,94.7000,206.50
rpm,23509.0,1144.849317,912.670971,124.10,489.40,856.00,1676.0000,4098.80
hours_since_maintenance,24042.0,172.630624,150.722469,0.00,42.87,121.61,295.5750,575.63
ambient_temp,24042.0,12.996398,2.883994,8.00,10.50,13.00,15.5000,18.00
rul_hours,24042.0,27.812510,26.393801,0.50,0.50,22.57,46.4100,98.34
failure_within_24h,24042.0,0.148074,0.355181,0.00,0.00,0.00,0.0000,1.00


# 10. Display Random Samples

Display a few random records to ensure that the dataset is loaded correctly and to inspect different observations.

In [36]:
df.sample(5)

,timestamp,machine_id,machine_type,vibration_rms,temperature_motor,current_phase_avg,pressure_level,rpm,operating_mode,hours_since_maintenance,ambient_temp,rul_hours,failure_within_24h,failure_type,estimated_repair_cost
15752,2024-01-02 16:16:00,14,Compressor,2.14,50.99,14.74,118.5,1795.5,normal,289.13,14.3,46.15,0,none,0
5329,2024-01-08 11:11:08,5,CNC,2.17,62.50,14.86,62.5,3019.2,normal,69.35,14.5,14.13,0,none,0
10097,2024-01-06 00:59:00,9,Pump,NaN,42.22,3.52,15.2,557.8,idle,330.25,15.6,0.50,0,none,0
22235,2024-01-07 08:16:19,19,Robotic Arm,0.91,45.99,5.73,47.8,352.2,normal,74.40,10.8,19.66,0,none,0
2417,2024-01-01 20:33:00,3,CNC,2.20,59.41,13.18,75.3,2964.8,normal,178.79,13.9,77.79,0,none,0


# 11. Analyze the Target Variable

The target variable for this project is **estimated_repair_cost**.

This section provides a statistical overview of the target variable to understand its distribution.

In [37]:
target = "estimated_repair_cost"

print(df[target].describe())

count    24042.000000
mean       608.870144
std       1566.793887
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max       7995.000000
Name: estimated_repair_cost, dtype: float64


# 12. Identify Numerical and Categorical Features

Separate the dataset into numerical and categorical features.

This helps in selecting appropriate preprocessing techniques during feature engineering.

In [38]:
numerical_columns = df.select_dtypes(include=["int64", "float64"]).columns.tolist()
categorical_columns = df.select_dtypes(include=["object"]).columns.tolist()

print(f"Numerical Features ({len(numerical_columns)}):")
print(numerical_columns)

print("\nCategorical Features ({})".format(len(categorical_columns)))
print(categorical_columns)

Numerical Features (11):
['machine_id', 'vibration_rms', 'temperature_motor', 'current_phase_avg', 'pressure_level', 'rpm', 'hours_since_maintenance', 'ambient_temp', 'rul_hours', 'failure_within_24h', 'estimated_repair_cost']

Categorical Features (4)
['timestamp', 'machine_type', 'operating_mode', 'failure_type']


# Key Observations

After completing the initial exploration, summarize the important findings from the dataset.

Example:

- Dataset contains 24042 rows and 15 columns.
- Target variable is **estimated_repair_cost**.
- The dataset contains both numerical and categorical features.
- Missing values: 4022
- Duplicate records: 0
- Dataset is suitable for regression analysis.